### PR TITLE
feat(#2069): Allows configure application from test

### PR DESCRIPTION
**Pull request by lordofthejars**

### CODE REVIEW

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/AbstractDelegatingConfigSource.java [hunk 1]
```diff
@@ -31,7 +31,7 @@ public AbstractDelegatingConfigSource(final ConfigSource delegate) {
      *
      * @return the delegate config source (not {@code null})
      */
-    protected @NotNull ConfigSource getDelegate() {
+    public @NotNull ConfigSource getDelegate() {
```
Comment (lordofthejars) [inline core/runtime/src/main/java/io/quarkus/runtime/configuration/AbstractDelegatingConfigSource.java]: I need to access to the delegate to check the test `ConfigSource` which contains setters method to add values programmatically. You'll see it later.

**Comment by lordofthejars** [comment]
When the PR is merged I'll send another one to `docs` repo with these new additions.

**Comment by lordofthejars** [comment]
Tests are running on my machine, so I need to figure out why not in CI env. I'll do it tomorrow.

**Comment by lordofthejars** [comment]
@stuartwdouglas @mkouba Hi Stuart I am facing some problems in CI system. I have implemented this feature, the tests are working locally, from IDE and also from Maven, but when They are run in CI system, the integration test is failing:  https://dev.azure.com/quarkus-ci/quarkus/_build/results?buildId=1643&view=logs&jobId=bfae5635-41b9-5c30-e2ee-d7864cc2a273&taskId=2c81e31e-52ec-544d-7911-5fc9d086633c&lineStart=17956&lineEnd=17957&colStart=1&colEnd=1

I suspect that there are some classpath issues when in it is CI system, but inspecting the azure script I do not see anything strange.

**Comment by lordofthejars** [comment]
If the problem is that in Azure test and Quarkus is running in different spaces, then I have another strategy that might work and it is creating a temporal file with the properties and add it to the classpath as a resource before starting Quarkus. But I'd like to know your point of view first.

File: test-framework/common/src/main/java/io/quarkus/test/common/configuration/TestConfigurationInitializer.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+package io.quarkus.test.common.configuration;
+
+import java.util.Map;
+
+/**
+ * +
+ * Interface that allows to set configuration properties at test runtime.
+ */
+public interface TestConfigurationInitializer {
+    Map<String, String> initilize();
```
Comment (Sanne) [inline test-framework/common/src/main/java/io/quarkus/test/common/configuration/TestConfigurationInitializer.java]: typo in the method name!

Also, some javadoc would be nice.

File: test-framework/common/src/main/java/io/quarkus/test/common/configuration/TestConfigurationInitializer.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.test.common.configuration;
+
+import java.util.Map;
+
+/**
+ * +
```
Comment (Sanne) [inline test-framework/common/src/main/java/io/quarkus/test/common/configuration/TestConfigurationInitializer.java]: `+` ?

**Comment by Sanne** [comment]
Hi @lordofthejars !

> If the problem is that in Azure test and Quarkus is running in different spaces

what do you mean by "spaces" ?

> I have another strategy that might work and it is creating a temporal file with the properties and add it to the classpath as a resource

I'm not a fan of that, it's slow and it kills any hope to eventually make it possible to run such tests in parallel. Also, if it works for you locally then I expect this to not be necessary on Azure either - please debug what's going on?

**Comment by lordofthejars** [comment]
@Sanne I cannot debug because it is working in my machine.

**Comment by Sanne** [comment]
ok, if you could at least rebase it, then others might be able to have a look too.

File: test-framework/common/src/main/java/io/quarkus/test/common/configuration/InMemoryConfigSource.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+/**
+ * In Memory Config Source that is used to store dynamically values from test.
+ */
+public class InMemoryConfigSource implements ConfigSource {
+
+    static final String NAME = "MemoryConfigSource";
+    private static final Map<String, String> PROPERTIES = new HashMap<>();
```
Comment (stuartwdouglas) [inline test-framework/common/src/main/java/io/quarkus/test/common/configuration/InMemoryConfigSource.java]: Why is this static, when none of the methods that access it are static?

File: test-framework/common/src/main/java/io/quarkus/test/common/configuration/InMemoryConfigSourcePopulator.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ * Class that populates configuration parameters from test to In Memory Config Source
+ */
+public final class InMemoryConfigSourcePopulator {
+
+    public void populate(final Class<?> testClass) {
+
+        final InMemoryConfigSource inMemoryConfigSource = findInMemoryConfigSource();
+        if (inMemoryConfigSource != null) {
+            final ConfigurationPropertiesExtractor configurationPropertiesExtractor = new ConfigurationPropertiesExtractor();
+            inMemoryConfigSource.addProperties(configurationPropertiesExtractor.extract(testClass));
```
Comment (stuartwdouglas) [inline test-framework/common/src/main/java/io/quarkus/test/common/configuration/InMemoryConfigSourcePopulator.java]: Are these config values ever cleared after the test? I can't seem to see where that happens

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: test-framework/common/src/main/java/io/quarkus/test/common/configuration/InMemoryConfigSource.java (hunk 4 above)
Comment (lordofthejars) [reply test-framework/common/src/main/java/io/quarkus/test/common/configuration/InMemoryConfigSource.java]: I cannot remember now, too long ago and I am getting old. But probably I need it for a test, then I changed the approach and I miss the `static`. Changed now.

File: test-framework/common/src/main/java/io/quarkus/test/common/configuration/InMemoryConfigSourcePopulator.java (hunk 5 above)
Comment (lordofthejars) [reply test-framework/common/src/main/java/io/quarkus/test/common/configuration/InMemoryConfigSourcePopulator.java]: True if the JVM is reused, going to fix that.

**Comment by lordofthejars** [comment]
@stuartwdouglas I have updated the PR. Now I have found the problem on why the tests are failing but the solution well I have not found it, I have been spending all day, but although it is because a class loader problem I cannot find how to fix it, probably it is really easy but now I cannot see it.

Look the thing is that in JUnit side after Quarkus is started I am getting the `InMemoryConfig` implementation to set the values. This config object is get from `appCl`:

https://github.com/quarkusio/quarkus/pull/2311/files#diff-985ca913dc620dcf2e54fd5e2b62a2bfR341

```
 final Iterable<ConfigSource> configSources = ConfigProvider.getConfig(classLoader).getConfigSources();
```

And then the configured values are set. The problem is that when `ConfigProvider.getConfig(classLoader)` is called from JUnit it returns one instance of `Config` object that it is different from the one is returned when Quarkus calls it to inject the values. I have tried to debug the Quarkus code but what I have seen is that Config object is retrieved from several  `ClassLoaders` and the one for the `InMemoryConfig` is different than `appCl` that is the application classloader returned in JUnit.

So now I am not sure how to proceed, I am not really advised on Quarkus bootstrap core so maybe you see clearly the problem and the solution and I can implement it.

Thanks.

**Comment by stuartwdouglas** [comment]
To be honest I would have probably just taken the simple approach of setting system properties before the test and restoring them afterwards.

**Comment by lordofthejars** [comment]
@stuartwdouglas Then you cannot run tests in parallel.

**Comment by lordofthejars** [comment]
@stuartwdouglas Do you think as the first approach that we sacrifice parallel tests (we documented as well) and we implemented as you said so at least we can have a base ground to improve later?

**Comment by FroMage** [comment]
@stuartwdouglas could you answer his question? This seems to be blocked.

@lordofthejars could you rebase, this has a conflict ATM.

**Comment by lordofthejars** [comment]
@FroMage I will do it when we choose the direction we need to follow so I do not resolve conflicts twice.

**Comment by emmanuelbernard** [comment]
Ping @stuartwdouglas

**Comment by lordofthejars** [comment]
Currently there is a workaround by implementing custom `QuarkusTestResourceLifecycleManager` and returning the new properties, but it is a workaround and not really improves the developer experience.

**Comment by stuartwdouglas** [comment]
I would just use a static approach, but with that said this is still somewhat problematic. Quarkus is only started once per test run with @QuarkusTest, so only the first test will have the full configuration applied, as a lot of the Quarkus properties are read at boot.

When running in JVM mode this will let you set user level config between tests, as long as it has not already been injected into application scoped beans. In substrate mode only the first config will be applied as the application is only started once.

Parallel tests are never going to work with this, as there is only ever a single running Quarkus app.

**Comment by lordofthejars** [comment]
@stuartwdouglas yeah this is far from perfect I agree.

**Comment by lordofthejars** [comment]
I think we can close this PR as everything can be done as a `QuarkusTestResource`

**Pull request closed without merging**